File: C++/NthTribonacci5.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int tribonacci(int n) {
        if(n<2)
            return n;
        
        int dp[] = {0,1,1};
                
        for(int i=3;i<=n;i++){
            dp[i%3] = dp[0] + dp[1] + dp[2];
        }
        return dp[n%3];
    }
};

int main()
{
    Solution sol;
    cout << sol.tribonacci(25);
    return 0;
}

File: JAVA/NthTribonacci5.java
//https://leetcode.com/problems/n-th-tribonacci-number/
import java.util.*;
class Solution1 {
    public int tribonacci(int n) 
    {
       return NthTribonacci(n,new HashMap<Integer,Integer>());
    }
    
    private int NthTribonacci(int n,HashMap<Integer,Integer> map)
    {   
        
        if(n==0 || n==1)
            return n;
        if(n==2)
            return n-1;
        
        int currentKey = n;
        
        if(map.containsKey(currentKey)){
            return map.get(currentKey);
        }
        
        int callOne   = NthTribonacci(n-1,map);
        int callTwo   = NthTribonacci(n-2,map);
        int callThree = NthTribonacci(n-3,map);
        
        map.put(currentKey,callOne+callTwo+callThree);
        
        return callOne+callTwo+callThree;
    }
}
class Tribonacci{
    public static void main(String[] args) {
        Solution1 sol = new Solution1();
        int n = 25;
        System.out.println(sol.tribonacci(n));
    }
}